src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs

```diff
@@ -1112,26 +1112,32 @@ private SymbolicRegexNode<TSet> PruneLowerPriorityThanNullability(SymbolicRegexB
 
                 case SymbolicRegexNodeKind.Concat:
                     Debug.Assert(_left is not null && _right is not null);
-                    //in a concatenation (X|Y)Z priority is given to XZ when X is nullable
-                    //observe that (X|Y)Z is equivalent to (XZ|YZ) and the branch YZ must be ignored
-                    //when X is nullable (observe that XZ is nullable because this=(X|Y)Z is nullable)
-                    //if X is not nullable then XZ is maintaned as is, and YZ is pruned
-                    //e.g. (a{0,5}?|abc|b+)c* reduces to c*
-                    //---
-                    //in a concatenation XZ where X is not an alternation, both X and Z are pruned
-                    //e.g. a{0,5}?b{0,5}? reduces to ()
-                    prunedNode = _left._kind == SymbolicRegexNodeKind.Alternate ?
-                        (_left._left!.IsNullableFor(context) ?
+                    prunedNode = _left._kind switch
+                    {
+                        // If the left side is a concatenation, then we bring it to right associative form to expose
+                        // the first element of the concatenation for the cases below.
+                        SymbolicRegexNodeKind.Concat => CreateConcat(builder, _left._left!, CreateConcat(builder, _left._right!, _right))
+                            .PruneLowerPriorityThanNullability(builder, context),
+                        // In a concatenation (X|Y)Z priority is given to XZ when X is nullable.
+                        // Observe that (X|Y)Z is equivalent to (XZ|YZ) and the branch YZ must be ignored
+                        // when X is nullable (observe that XZ is nullable because this=(X|Y)Z is nullable).
+                        // If X is not nullable then XZ is maintaned as is, and YZ is pruned. For example,
+                        // (a{0,5}?|abc|b+)c* reduces to c*.
```

Comment:
To check my understanding, this is only true because:
- `a{0,5}?` is lazy with a lower bound of 0 and thus will first match empty
- `c*` also matches empty

If either of those wasn't true, this reduction would not be possible, yes?  Which of X,Y,Z in your prose aligns with the `c*`?

For example, if the `c*` were instead `c+`, then matching the initial expression against `ac` would require matching the `a` against the first iteration of the `a` from the lazy loop.

Reply:
The general rule is that in `WZ` if `W` is **high-priority-nullable** (e.g. starts with a lazy loop with lower bound 0 as the first alternative in `W`) then the other alternatives in `W` do not matter and are pruned.

Reply:
I'm still not understanding. Let's say I have the pattern:
`(a*?|b)c`
I believe what you're saying is the b here can be removed (the alternation is nullable and it starts with a lazy loop with a lower bound of zero). But changing this to `a*?c` changes semantics. If I match against the input "bc", the original would match "bc" whereas the pruned would match only "c". Have I misunderstood?

Reply:
The expression `(a*?|b)c` is not nullable, the **precondition** is that `WZ` is nullable, meaning `Z` is also nullable, but in your example `c` is not nullable. Above `Z` is some nullable regex, e.g.. `(bc)*`.

Reply:
> the precondition is that WZ is nullable

Ok